[tide] Allow PRs to be picked for batch while still pending tests

arturo-skydio opened the pull request:
Allows tide to select PRs without the need for a first successful presubmit(s) build, this is intended to increase the likelihood of batch PRs. This logic disqualifies any failed individual builds and just include pending or passing.

Primary usecase is where teams have quick fixes that can be added to batch while still pending their individual builds (where build times are longer than usual).

Introducing a boolean config `batchAllowPending`, if set to `true` while enable this functionality.

petr-muller commented:
/ok-to-test

The code looks straightforward, but what's the use case? If the PR is failing tests what's the point of including in a batch?

arturo-skydio commented:
> /ok-to-test
> 
> The code looks straightforward, but what's the use case? If the PR is failing tests what's the point of including in a batch?

Hi, thanks for reviewing, added some description to the PR, not sure if this usecase applies to other teams using Prow but for us it was something that helped with our builds being around 2 hours

petr-muller commented on prow/tide/tide.go:
```diff
@@ -1011,9 +1031,13 @@ func (c *Controller) pickBatch(sp subpool, cc map[int]contextChecker) ([]PullReq
 	// we must choose the oldest PRs for the batch
 	sort.Slice(sp.prs, func(i, j int) bool { return sp.prs[i].Number < sp.prs[j].Number })
 
+	batchSkipTests := false
+	if c.config().Tide.BatchSkipTests != nil && *c.config().Tide.BatchSkipTests {
+		batchSkipTests = true
```
```suggestion
	var batchSkipTests bool
	if c.config().Tide.BatchSkipTests != nil {
		batchSkipTests = *c.config().Tide.BatchSkipTests 
```

is IMO a bit more idiomatic and less wordy, but I would not even be afraid of doing this in a single line (`bst := (ptr != nil && *ptr`)

arturo-skydio replied:
will try the one liner thanks!

petr-muller commented on prow/tide/tide.go:
```diff
@@ -1011,9 +1031,13 @@ func (c *Controller) pickBatch(sp subpool, cc map[int]contextChecker) ([]PullReq
 	var candidates []PullRequest
 	for _, pr := range sp.prs {
-		if isPassingTests(sp.log, c.ghc, pr, cc[int(pr.Number)]) {
+		if (batchSkipTests && isPendingTests(sp.log, c.ghc, pr, cc[int(pr.Number)])) || isPassingTests(sp.log, c.ghc, pr, cc[int(pr.Number)]) {
```
I think that requiring all checks to be pending may be too strict when there are jobs that run very fast. I think you'd want all checks to be either passing or pending?

arturo-skydio replied:
As far as I understand it I'm only checking to see if there are failing tests because I'm looping over unsuccessfulContext pr's which includes failures and pending, so maybe whats needed here is a better function name like `failingTests` ?

arturo-skydio replied:
Renamed to `isFailingTests` , let me know if that works

petr-muller replied:
Yeah, thanks for explanation, I misread the function :)

arturo-skydio commented:
@petr-muller sorry, just renamed `batchSkipTests` to `batchAllowPending` and added docs that are more in line with actual behavior, can you reapprove? Thanks again

arturo-skydio commented:
/assign @cblecker

arturo-skydio commented:
Any feedback here? Or suggestions on how to get this merged?
